ArduCopter/land_detector.cpp

```diff
@@ -156,13 +156,9 @@ void Copter::set_land_complete(bool b)
 
     // tell AHRS flying state
     set_likely_flying(!b);
-    
-    // trigger disarm-on-land if configured
-    bool disarm_on_land_configured = (g.throttle_behavior & THR_BEHAVE_DISARM_ON_LAND_DETECT) != 0;
-    const bool mode_disarms_on_land = flightmode->allows_arming(AP_Arming::Method::LANDING) && !flightmode->has_manual_throttle();
 
-    if (ap.land_complete && motors->armed() && disarm_on_land_configured && mode_disarms_on_land) {
-        arming.disarm(AP_Arming::Method::LANDED);
+    if (b) {
+        flightmode->land_complete();
     }
```

Comment:
```suggestion
    if (!b) {
        // Not landed
        return;
    }
    
    // trigger disarm-on-land if configured
    if ((g.throttle_behavior & THR_BEHAVE_DISARM_ON_LAND_DETECT) == 0) {
        // not configured to disarm
        return;
    }

    if (!motors->armed()) {
        // not armed
        return;
    }

    if (has_manual_throttle()) {
        return;
    }
    // we should probably have a "allows_disarm(...)" callback
    if (!allows_arming(AP_Arming::Method::LANDING)) {
        return;
    }

    AP::arming().disarm(AP_Arming::Method::LANDED);
```

Reply:
Yep, rewrote the thing in-place instead.